File: src/05smallest_multiple.js
let getPrimeFactors = (n) => {
	
	let factors = {}

	let addFactor = (n) => {
		if(factors[n] === undefined){
			factors[n] = 0;
		}
		factors[n]++;
	}

	while(n % 2 === 0){
		n = n / 2;
		addFactor(2);
	}

	for(let i = 3; i <= Math.sqrt(n); i += 2){
		while(n % i === 0){
			n = n / i;
			addFactor(i);
		}
	}

	if(n > 1){
		addFactor(n);
	}

	return factors;
}

let smallestMultiple = (limit) => {

	let factors = {};

	for(let i = 2; i <= limit; i++){
		let newFactors = getPrimeFactors(i);

		for(let factor in newFactors){
			if(!factors[factor] || factors[factor] < newFactors[factor]){
				factors[factor] = newFactors[factor];
			}
		}
	}

	return Object.keys(factors).reduce((product, key) => {
		return product * Math.pow(key, factors[key]);
	}, 1);
}

File: python/05smallest_multiple.py
#!/usr/bin/env python -tt
import sys

def get_factors(n):
	factors = {}
	d = 2
	while n > 1:
		while n % d == 0:
			n /= d
			if d in factors:
				factors[d] += 1
			else:
				factors[d] = 1
		d += 1
	return factors


def smallest_multiple(n):
	
	factor_hash = {}

	for i in range(0, n):
		i_hash = get_factors(i)

		for factor in i_hash:
			if factor in factor_hash:
				factor_hash[factor] = factor_hash[factor] if (factor_hash[factor] > i_hash[factor]) else i_hash[factor]
			else:
				factor_hash[factor] = i_hash[factor]

	print factor_hash

	product = 1

	for factor in factor_hash:
		product = product * (factor**factor_hash[factor])

	return product



def main():
  n = sys.argv[1]
  print smallest_multiple(int(n))

if __name__ == "__main__":
	main()